
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://mdu-phl.github.io/tbtamr/usage/">
      
      
        <link rel="prev" href="../installation/">
      
      
        <link rel="next" href="../customisation/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Usage - tbtAMR</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#before-you-start" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="tbtAMR" class="md-header__button md-logo" aria-label="tbtAMR" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            tbtAMR
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/mdu-phl/tbtamr" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    mdu-phl/tbtamr
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="tbtAMR" class="md-nav__button md-logo" aria-label="tbtAMR" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    tbtAMR
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/mdu-phl/tbtamr" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    mdu-phl/tbtamr
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#before-you-start" class="md-nav__link">
    <span class="md-ellipsis">
      Before you start
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Before you start">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#is-your-vcf-file-correctly-formatted" class="md-nav__link">
    <span class="md-ellipsis">
      Is your vcf file correctly formatted?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#have-you-installed-tbtamr-to-run-the-way-you-want-to-run" class="md-nav__link">
    <span class="md-ellipsis">
      Have you installed tbtAMR to run the way you want to run?
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Have you installed tbtAMR to run the way you want to run?">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-note-on-lineage-calling" class="md-nav__link">
    <span class="md-ellipsis">
      A note on lineage calling.
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generating-genomic-dst-using-default-criteria-tbtamr-defaults" class="md-nav__link">
    <span class="md-ellipsis">
      Generating genomic DST using default criteria (tbtAMR defaults)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Generating genomic DST using default criteria (tbtAMR defaults)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#vcf-input" class="md-nav__link">
    <span class="md-ellipsis">
      VCF input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paired-end-reads-input" class="md-nav__link">
    <span class="md-ellipsis">
      paired-end reads input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-phylogenetic-lineage-to-outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Add phylogenetic lineage to outputs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generating-genomic-dst-using-custom-criteria" class="md-nav__link">
    <span class="md-ellipsis">
      Generating genomic DST using custom criteria
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#use-tbtamr-to-generate-annotated-vcf-file-no-dst" class="md-nav__link">
    <span class="md-ellipsis">
      Use tbtAMR to generate annotated VCF file (no DST)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#search-your-catalogue-for-variant-information" class="md-nav__link">
    <span class="md-ellipsis">
      Search your catalogue for variant information
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      Output files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../customisation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Customisation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../cascade_reporting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cascade reporting
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#before-you-start" class="md-nav__link">
    <span class="md-ellipsis">
      Before you start
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Before you start">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#is-your-vcf-file-correctly-formatted" class="md-nav__link">
    <span class="md-ellipsis">
      Is your vcf file correctly formatted?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#have-you-installed-tbtamr-to-run-the-way-you-want-to-run" class="md-nav__link">
    <span class="md-ellipsis">
      Have you installed tbtAMR to run the way you want to run?
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Have you installed tbtAMR to run the way you want to run?">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-note-on-lineage-calling" class="md-nav__link">
    <span class="md-ellipsis">
      A note on lineage calling.
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generating-genomic-dst-using-default-criteria-tbtamr-defaults" class="md-nav__link">
    <span class="md-ellipsis">
      Generating genomic DST using default criteria (tbtAMR defaults)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Generating genomic DST using default criteria (tbtAMR defaults)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#vcf-input" class="md-nav__link">
    <span class="md-ellipsis">
      VCF input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paired-end-reads-input" class="md-nav__link">
    <span class="md-ellipsis">
      paired-end reads input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-phylogenetic-lineage-to-outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Add phylogenetic lineage to outputs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generating-genomic-dst-using-custom-criteria" class="md-nav__link">
    <span class="md-ellipsis">
      Generating genomic DST using custom criteria
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#use-tbtamr-to-generate-annotated-vcf-file-no-dst" class="md-nav__link">
    <span class="md-ellipsis">
      Use tbtAMR to generate annotated VCF file (no DST)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#search-your-catalogue-for-variant-information" class="md-nav__link">
    <span class="md-ellipsis">
      Search your catalogue for variant information
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      Output files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Usage</h1>

<h2 id="before-you-start">Before you start</h2>
<p><code>tbtAMR</code> can take any correctly formatted VCF file - which means that it is sequence platform agnostic. Alternatively, you can also supply Illumina paired-end reads as sqeuence input, providing you have installed <code>mutAMR</code> correctly. However when designing and developing <code>tbtAMR</code> some assumptions were made. </p>
<ul>
<li>If you are using vcf as input - it is formatted correctly.</li>
<li>Annotation was undertaken using snpEff or another annotation tool which adheres to specification found <a href="https://pcingola.github.io/SnpEff/adds/VCFannotationformat_v1.0.pdf">here</a>. If you supply an un-annotated vcf and have snpEff installed, <code>tbtAMR</code> will annotate for you (see HERE for installation instructions)</li>
</ul>
<h3 id="is-your-vcf-file-correctly-formatted">Is your vcf file correctly formatted?</h3>
<p>There is a standard for generation of VCF files, you can learn more about this <a href="http://samtools.github.io/hts-specs/">here</a>. Most bioinformatics tools will (should) follow this standard, if the tool you are using does not - there is no guarantee that <code>tbtAMR</code> behaviour will be as expected. Basic checks will be undertaken to ensure that your VCF is correct and if obvious issues are detected <code>tbtAMR</code> will error and ask you to check your inputs. However, if you are using a tool that does not follow this standard, but generates a file that mostly looks correct, <code>tbtAMR</code> may not detect the difference, it will appear to run correctly but the outputs may be unexpected or incorrect. If this is the case, please check your VCF generation and make sure that it follows the above linked specifications.</p>
<h3 id="have-you-installed-tbtamr-to-run-the-way-you-want-to-run">Have you installed <code>tbtAMR</code> to run the way you want to run?</h3>
<p>If you are supplying an annotated VCF as input and have no need for VCF generation or annotation, you can just install <code>tbtAMR</code> as describe <a href="https://github.com/MDU-PHL/tbtamr/wiki/Installation#pip-simple-install">here</a>. However, if you would like to supply paired-end fastq files as input, you will need to make sure that you have installed <code>tbtAMR</code> as descibed <a href="https://github.com/MDU-PHL/tbtamr/wiki/Installation#bioconda-recommended">here</a>.</p>
<p>Alternatively, you can also just install <code>snpEff</code> and <code>tbtAMR</code> and described <a href="https://github.com/MDU-PHL/tbtamr/wiki/Installation#tbtamr--annotation">here</a> if you want <code>tbtAMR</code> to annotate your VCF file.</p>
<h4 id="a-note-on-lineage-calling">A note on lineage calling.</h4>
<p>By default, <code>tbtAMR</code> does not call lineage for <em>M. tuberculosis</em>, you can do this if you like by following the installation instructions <a href="https://github.com/MDU-PHL/tbtamr/wiki/Installation#tbtamr-and-lineage-calling">here</a> and using the <code>--call_lineage</code> flag.</p>
<h2 id="generating-genomic-dst-using-default-criteria-tbtamr-defaults">Generating genomic DST using default criteria (tbtAMR defaults)</h2>
<h3 id="vcf-input">VCF input</h3>
<div class="highlight"><pre><span></span><code>tbtamr predict --vcf sample.vcf.gz -s sample_name
</code></pre></div>
<h3 id="paired-end-reads-input">paired-end reads input</h3>
<div class="highlight"><pre><span></span><code>tbtamr full -1 R1.fq.gz -2 R1.fq.gz -s sample_name
</code></pre></div>
<h3 id="add-phylogenetic-lineage-to-outputs">Add phylogenetic lineage to outputs</h3>
<p>Please make sure that you have installed <code>pathogen-profiler</code> as described <a href="https://github.com/MDU-PHL/tbtamr/wiki/Installation#tbtamr-and-lineage-calling">here</a>.</p>
<p><div class="highlight"><pre><span></span><code>tbtamr predict --vcf sample.vcf.gz -s sample_name --call_lineage
</code></pre></div>
OR</p>
<div class="highlight"><pre><span></span><code>tbtamr full -1 R1.fq.gz -2 R1.fq.gz -s sample_name --call_lineage
</code></pre></div>
<h2 id="generating-genomic-dst-using-custom-criteria">Generating genomic DST using custom criteria</h2>
<p>Check out the instructions for defining criteria <a href="https://github.com/MDU-PHL/tbtamr/wiki/Customising-your-inputs">here</a>, example files can also be found <a href="https://github.com/MDU-PHL/tbtamr/blob/master/example_criteria">here</a>.</p>
<p><div class="highlight"><pre><span></span><code>tbtamr predict --vcf sample.vcf.gz -s sample_name -c custom_mutational_catalogue.csv -cfg custom_db_config.json -r custom_interpretation_criteria.csv -cr custom_classification_criteria.csv
</code></pre></div>
OR</p>
<div class="highlight"><pre><span></span><code>tbtamr full -1 R1.fq.gz -2 R1.fq.gz -s sample_name -c custom_mutational_catalogue.csv -cfg custom_db_config.json -r custom_interpretation_criteria.csv -cr custom_classification_criteria.csv
</code></pre></div>
<h2 id="use-tbtamr-to-generate-annotated-vcf-file-no-dst">Use <code>tbtAMR</code> to generate annotated VCF file (no DST)</h2>
<p>It is unclear why you may want to do this - but the functionality is here for you if you want to use it</p>
<p><div class="highlight"><pre><span></span><code>tbtamr fq2vcf -1 R1.fq.gz -2 R1.fq.gz -s sample_name
</code></pre></div>
OR
<div class="highlight"><pre><span></span><code>tbtamr annotate --vcf sample.vcf.gz -s sample_name
</code></pre></div></p>
<h2 id="search-your-catalogue-for-variant-information">Search your catalogue for variant information</h2>
<p><code>tbtAMR</code> also allows you to quickly peruse your catalogue - which can be useful if you don't want to break excel and you have a large catalogue.</p>
<p><div class="highlight"><pre><span></span><code>tbtamr search -q rpoB_p.Ser540Leu rpoB_p.Ser540Ala
</code></pre></div>
If you are using a custom catalogue - you will need to do </p>
<div class="highlight"><pre><span></span><code>tbtamr search -q some_query -c custom_mutational_catalogue.csv -cfg custom_db_config.json
</code></pre></div>
<h2 id="output-files">Output files</h2>
<p><code>tbtAMR</code> will output two files as standard. </p>
<ol>
<li>
<p>The first is <code>tbtamr_results.csv</code>. This is essentially the raw output of your interpretative criteria, this may include reportable and non-reportable results (depending on how the interpretative criteria and catalogue configuration are set up).</p>
</li>
<li>
<p>The second file is <code>tbtamr_linelist_report.csv</code>. This file is a filtered version of what is in the <code>tbtamr_results.csv</code>, with only results deemed appropriate for reporting (as defined by the interpretative criteria and catalogue configuration).</p>
</li>
</ol>
<p>If you have added the <code>--cascade</code> flag, and have <code>cascade_reporting</code> set up in the catalogue configuration, a 3rd file will be created
3. <code>tbtamr_linelist_cascade_report.csv</code>, this is a further filtered version of the linelist report, with specific drugs reported based on the resistance profile detected. For more information see <a href="https://github.com/MDU-PHL/tbtamr/wiki/Cascade-reporting">here</a> and <a href="https://github.com/MDU-PHL/tbtamr/wiki/Customising-your-inputs#cascade-reporting">here</a>.</p>
<p>Note all output files are designed to be simple <strong>csv</strong> to allow for straightforward management and incorporation into LIMS. <code>tbtAMR</code> does not make any assumptions as to how the reports should appear or be communicated.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/mdu-phl/tbtamr" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>